src/lib.rs

```diff
@@ -0,0 +1,127 @@
+#[deny(clippy::all)]
+///! Graph API Traits
+///
+use rand::SeedableRng;
+
+#[allow(dead_code)]
+/// A graph layer name.
+struct Layer(&'static str);
+
+/// Abstract object in a graph, eg. node or edge.
+trait GraphObject {
+    /// Identifier of all graph objects.
+    type Id;
+
+    /// Local graph object data.
+    type Data;
+
+    /// Return the edge id.
+    fn id(&self) -> Self::Id;
+
+    /// Get local data.
+    fn data(&self) -> &Self::Data;
+
+    /// Get mutable local data.
+    fn data_mut(&self) -> &mut Self::Data;
+}
+
+/// A graph  node.
+trait Node: GraphObject {
+    /// The edge type connected to this node.
+    type Edge: GraphObject;
+
+    /// Get connected nodes.
+    fn neighbors(&self) -> Iterator<Item = &Self>;
+
+    /// Get incoming and outgoing edges.
+    fn edges(&self) -> Iterator<Item = &Self::Edge>;
+}
+
+/// A graph edge between two nodes.
+trait Edge: GraphObject {
+    /// Get the edge weight.
+    fn weight(&self) -> f64;
+
+    /// The object from which the edge starts.
+    fn from(&self) -> <Self as GraphObject>::Id;
+
+    /// The object to which the edge points.
+    fn to(&self) -> <Self as GraphObject>::Id;
+}
+
+/// The Graph API
+trait GraphAPI<G: Graph> {
+    /// Entropy used to seed the CPRNG.
+    type Entropy;
+
+    /// Add a graph layer.
+    fn add_layer(&mut self, layer: Layer);
+
+    /// Remove a graph layer. The layer must not contain any nodes.
+    fn remove_layer(&mut self, layer: &Layer);
```

Comment:
Total bikeshed: if you _move_ the `Layer` when you call `add_layer`, how are you supposed to pass a reference to it when you remove it? Perhaps `add_layer` should take the reference and `remove_layer` move it? 🤔

Reply:
You reconstruct it or store a copy: `self.remove_layer(&Layer("osrank"))`